### ISSUE TITLE
BS4/Bootstrap-Vue Accessibility recommendations

### ISSUE THREAD

**Issue by tmorehouse** [bootstrap-vue/bootstrap-vue#333]
Bootstrap V4 is addressing some of the accessibility issues: https://github.com/twbs/bootstrap/issues/22549, Of which bootstrap-vue has started to already address.

Here is  little list of items that are complete or need work for aria and/or keyboard navigation (to be updated as we progress through them):

**Components - ARIA / Keyboard Navigation**
- [x] `b-form-radio` input wrapped in label (which negates need for `for` on label)
- [x] `b-form-checkbox` input wrapped in label (which negates need for `for` on label)
- [x] `b-progress` ARIA `progress` attributes
- [x] `b-modal` attributes `aria-labelledby` and `aria-describedby`. Section `roles` & semantic elements. (PR #247)
- [x] `b-modal` Focus first input/button on open (PR #247). Allow page author to specify input to be focused when opened via `@shown` event (PR #378)
- [x] `b-modal` focus the modal content when opened
- [x] `b-tabs` & `b-tab` attributes `aria-controls` and `aria-labelelledby`. Section `roles` (PR #339)
- [x] `b-tabs` & `b-tab` keyboard navigation (PR #339)
- [x] `b-alert` has aria `role="alert"`,  `aria-live` and `aria-atomic` (PR #340)
- [x] `b-form-file` Needs focus styling in custom-file-input mode for keyboard users (PR #1033)
- [x] `b-form-fieldset` attribute `for` when ID supplied on `,form-control`. Has `role="alert"`, `aria-live`, `aria-atomic` on feedback. and `aria-describedby` (PR #340)
- [x] `b-breadcrumb` ~should have role of navigation (PR #340)~ Add `aria-current` to active crumb (PR #526)
- [x] `b-input-group` needs `role="group"` (PR #340)
- [x] `b-button-group` `role="group"`, make new toolbar component  (PR #367)
- [x] `b-button-toolbar` `role="toolbar"` + optional keyboard navigation (PR #367)
- [x] `b-nav` & `b-nav-link` aria attributes (PR #358)
- [x] `b-dropdown` keyboard navigation (PR #274)
- [x] `b-nav-dropdown` `aria-` attributes (PR #358)
- [x] `b-nav-dropdown` open on ENTER or SPACE (see Issue #348) (PR #349)
- [x] `b-nav-toggle` `aria-` attributes  (PR #358, #410, #411, #412)
- [x] ~~`b-nav` & `b-nav-link` keyboard navigation~~ *(leave as standard TAB key navigation)*
- [x] `b-collapse` Reflect expanded/collapsed state on trigger element (PR #358 & #519)
- [x] `b-pagination` `aria-[label|current|setsize|posinset]` attributes.  (Note: `aria-controls` should be set manually in the `<b-pagination>`markup) (PR #364)
- [x] `b-pagination` keyboard navigation (PR #364 & #377)
- [x] `b-carousel` `aria-[controls|label|current|setsize|posinset]` attributes and `roles`. (PR #380)
- [x] `b-carousel` keyboard left/right control (PR #380, #420)
- [x] `b-carousel` Pause slide scrolling on `mouseenter` and resume on `mouseout`.
- [x] `b-tooltip` `aria-` attributes  (aria-live, aria-describedby, etc
- [ ] `b-popover` `aria-` attributes  (aria-live, aria-describedby, etc. Also needs ability to change `role` from `tooltip` to `popover` or `dialog`
  - Popovers can have some issues with regards to accessibility, especially for interactive popover content, and possibly should have a role other than `tooltip` (configurable). https://github.com/twbs/bootstrap/issues/18618

**Directives**
- [x] `v-b-modal` return focus to open trigger element.
- [x] `v-b-toggle` attributes `aria-controls` and `aria-expanded`. (PR #519)

**Documentation**
- [x] Create ARIA best practices for Bootstrap-Vue - sections added to each compoent
- [x] Add documentation where needed

**Comment by tmorehouse** [comment]
Also found a discussion on autofocus and modal open:

https://github.com/twbs/bootstrap/issues/22402

Which brings up a good point about autofocusing any `input`/`textarea` when on a mobile device with virtual keyboard.  I might revisit the autofocus selectors in `modal.vue` to only look for `buttons`, `select` or the `[autofocus]:first` attribute, and if none found, then just focus the `.modal-content`

**Comment by tmorehouse** [comment]
Here is  little list of items that are complete or need work for aria and/or keyboard navigation (I'll update as I progress through them):

**Components & ARIA / Keyboard Navigation**
- [x] `b-modal` aria reference attributes
- [ ] `b-form-fieldset` has attribute `for` when ID supplied on `b-input` & `b-select`. has proper `role="alert"` and `aria-live` for form-control-feedback, but needs  `aria-describedby` referencing feedback element or `aria-live` + `aria-atomic` on fieldset
- [x] `b-form-radio` input wrapped in label (which negates need for `for` on label)
- [x] `b-progress` aria attributes
- [x] `b-alert` aria `role="alert"`, but should possibly have `aria-live` and `aria-atomic` set as well (for dynamic content, specifically for alerts that appear from hidden state)
- [ ] `b-modal` only focus first non keyboard input on open (for better mobile support)
- [x] `b-tabs` & `b-tab` aria attributes (PR coming soon)
- [x] `b-tabs` & `b-tab` keyboard navigation (PR coming soon)
- [ ] `b-button-group` toolbar variant keyboard navigation
- [ ] `b-input-group` needs `role="group"`
- [ ] `b-nav` & `b-nav-link` aria attributes
- [ ] `b-nav` & `b-nav-link` keyboard navigation
- [ ] `b-nav-dropdown` `aria-` attributes
- [ ] `b-nav-toggle` `aria-` attributes
- [ ] `b-b-pagination` `aria-` attributes [`selected`, `setsize`, `posinset`, `controls` *(optional)*]
- [ ] `b-crousel` `aria-` attributes

**Directives**
- [x] `v-b-modal` return focus to open trigger element.
- [ ] `v-b-toggle` should have `aria-controls` (on button) and `aria-expanded` (on either button or target)

For components that need `aria-` attributes that reference IDs, should we force the user to supply unique ID? Or dynamically remove `aria-` attributes that need IDs by setting them to null (breaking some screen reader usability)? *(caveat emptor)*

**Documentation**
- [ ] Create ARIA best practices for Bootstrap-Vue
- [ ] Create examples for ARIA component cross referencing

**Comment by tmorehouse** [comment]
Partway through the list via a couple of PRs.

**Comment by Decrepidos** [comment]
Nice work. If I find any isdues in regards to accessibility I will post my findings. 😁

**Comment by tmorehouse** [comment]
The list (https://github.com/bootstrap-vue/bootstrap-vue/issues/333#issuecomment-299114395) is getting shorter :)

**Comment by tmorehouse** [comment]
With respect to ID requirements, I am now leaning more towards the page author providing a unique Id on each component, and the component taking advantage of the provided ID to generate sub IDs as needed., and moving away from auto generated IDs (removing requirement for `mixins/generate-id`), with the possibility of logging warning messages (in devMode only) that certain components are missing IDs

**Comment by pi0** [comment]
Nice work so far :) Did you notice ARIA problems with NavBar ? (Tested using [this](https://chrome.google.com/webstore/detail/accessibility-developer-t) chrome extension)

![image](https://cloud.githubusercontent.com/assets/5158436/26169367/91b7126a-3b52-11e7-8391-ecf3ce559e91.png)

**Comment by tmorehouse** [comment]
@pio I'll take a look and make sure the Attributes are correct with Navbar.  Was this using the doc fiddle for NavBar?

-Troy

**Comment by pi0** [comment]
Troy, Screenshot above was from docs main page. This may be just because of incorrect usage in docs :))

**Comment by tmorehouse** [comment]
OK, I'll check both.

**Comment by tmorehouse** [comment]
Is docs using latest release or master?

**Comment by pi0** [comment]
Latest. I'm going to release 0.16.0 in five minutes :))

**Comment by tmorehouse** [comment]
OK I'll wait until next release to check.

**Comment by pi0** [comment]
Released. Still having same problems. (Here is [nav.vue](https://github.com/bootstrap-vue/bootstrap-vue/blob/v0.16.0/docs/nuxt/includes/nav.vue) implementation)

**Comment by tmorehouse** [comment]
Ah, I see that `b-collapse` is not rendering the ID on it's root div.  I'll create a quick PR for this.

**Comment by tmorehouse** [comment]
@pi0 PR #410

**Comment by tmorehouse** [comment]
Although I see that `aria-expanded` is set to `false` and it should be `true` in this case (since it isn't collapsed in wide view port).

Hmmm might need to find a way to get the `aria-expanded` to reflect the display based on if the collapse is visible (maybe by checking the computed Display property, or checking the visibility of the element (i.e. does it have a position in document).

**Comment by tmorehouse** [comment]
PR #411 fixes where the `aria-expanded` attribute appears for nav-collapse

**Comment by tmorehouse** [comment]
`carousel` may have a small bug when clicking the indicator buttons during a transition. I'll look into this tomorrow.

**Comment by tmorehouse** [comment]
PR #449 removed the `generate-id` mixin requirements for form input comonents.  Users, if they want ARIA compliance, will need to set an ID prop on the components.

**Comment by tmorehouse** [comment]
PR #519 adds ARIA attributes to `v-b-toggle` directive

**Comment by Decrepidos** [comment]
I dissagree that b-breadcrumb should have role of navigation. I think the supported aria-current=location is sufficient. A role nav or HTML5 native <nav> is overkill for breadcrumbs.

**Comment by tmorehouse** [comment]
@dleesalestreamsoft interesting point. Maybe allowing the user to change the role as a prop would suffice, depending on use case, And a prop that can be used to select semantic element is used as a wrapper.

I'll look into `aria-current` attribute on `b-breadcrumb`

**Comment by Decrepidos** [comment]
Hi Troy

I used to wrap breadcrumbs in a role region with aria-label=breadcrumbs.
Since aria-current now has widespread support I have adopted the approach
of aria-current=page for main navigation links and
aria-current=location for breadcrumbs and dropped the region and label. I
had a discussion on the same subject recently with a blind a11y developer
who suggested this approach.

Best regards
Laurence

**Comment by tmorehouse** [comment]
@Decrepidos what about using `aria-current="page"` instead of `aria-current="location"`?

**Comment by Decrepidos** [comment]
You can use either. Both are correct. This is a useful article on aria-current 
https://tink.uk/using-the-aria-current-attribute/

**Comment by Decrepidos** [comment]
I notice you posted the link. Sorry can't seem to edit post on the phone 🤔

**Comment by tmorehouse** [comment]
@Decrepidos I've defaulted it to `location` but have created a prop where the user can change it to another value such as `page` or `true` in PR #526

**Comment by Decrepidos** [comment]
@tmorehouse to clarify my personal view is page for navigation. Location for breadcrumbs and true for elements like carousel slide buttons. 😁

**Comment by Decrepidos** [comment]
Perfect solution 😁

**Comment by tvld** [comment]
Google Lighthouse for PWA,  also giving:
```
[aria-*] attributes are not valid or misspelled.
Assistive technologies, like screen readers, can't interpret ARIA attributes with invalid names. Learn more.
View failing elements
```
```html
<div role="listitem" 
    class="carousel-item active carousel-item-left"
    aria-current="true"
    aria-posinset="2" 
    aria-setsize="4" 
    tabindex="-1" 
    aria-controlledby="carousel1___BV_indicator_2_"
    aria-hidden="false">
```

**Comment by tmorehouse** [comment]
Some validators will complain about `aria-current`.  Which attribute is it complaining about?

**Comment by tvld** [comment]
Unfortunately, it does not say which one, but ``aria-controlledby`` is the only one not found on https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques

**Comment by Decrepidos** [comment]
Thats because it dies not exist. I suggest it should be aria-controls

**Comment by tmorehouse** [comment]
Hmm, I'll look into that one.

**Comment by tmorehouse** [comment]
Should be fixed in next release via PR #1448

**Comment by tmorehouse** [comment]
@tvld & @Decrepidos this has been fixed in release v1.4.0

**Comment by Decrepidos** [comment]
Nice work
